File: windows/src/Conduit.Core/Protocol/Packet.cs
using System.Text.Json;
using System.Text.Json.Nodes;

namespace Conduit.Core.Protocol;

/// <summary>Canonical list of packet <c>type</c> values. Must match PROTOCOL.md and the Android side.</summary>
public static class PacketType
{
    public const string Identity = "identity";
    public const string PairRequest = "pair-request";
    public const string PairResponse = "pair-response";
    public const string Ping = "ping";
    public const string Pong = "pong";
    public const string Clipboard = "clipboard";
    public const string FileOffer = "file-offer";
    public const string FileChunk = "file-chunk";
    public const string FileComplete = "file-complete";
    public const string Notification = "notification";
    public const string NotificationAction = "notification-action";
    public const string MediaState = "media-state";
    public const string MediaCommand = "media-command";
    public const string RemoteCommand = "remote-command";
    public const string Battery = "battery";
    public const string DeviceStatus = "device-status";
    public const string SmsList = "sms-list";
    public const string SmsSend = "sms-send";
    public const string Error = "error";
}

/// <summary>
/// The envelope every message travels in: { id, type, ts, body }. The body is an
/// arbitrary JSON object accessed via the typed helpers below.
/// </summary>
public sealed class Packet
{
    public string Id { get; init; } = Guid.NewGuid().ToString("N");
    public required string Type { get; init; }
    public long Ts { get; init; } = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
    public JsonObject Body { get; init; } = new();

    private static readonly JsonSerializerOptions JsonOpts = new()
    {
        WriteIndented = false
    };

    public static Packet Create(string type, Action<JsonObject>? build = null)
    {
        var body = new JsonObject();
        build?.Invoke(body);
        return new Packet { Type = type, Body = body };
    }

    public string? GetString(string key) => Body[key]?.GetValue<string>();
    public long GetLong(string key, long fallback = 0) =>
        Body[key] is { } n && n.AsValue().TryGetValue<long>(out var v) ? v : fallback;
    public int GetInt(string key, int fallback = 0) => (int)GetLong(key, fallback);
    public bool GetBool(string key, bool fallback = false) =>
        Body[key] is { } n && n.AsValue().TryGetValue<bool>(out var v) ? v : fallback;

    public string ToJson()
    {
        var obj = new JsonObject
        {
            ["id"] = Id,
            ["type"] = Type,
            ["ts"] = Ts,
            ["body"] = Body.DeepClone()
        };
        return obj.ToJsonString(JsonOpts);
    }

    public static Packet FromJson(string json)
    {
        var node = JsonNode.Parse(json)?.AsObject()
                   ?? throw new FormatException("Packet JSON is not an object");
        return new Packet
        {
            Id = node["id"]?.GetValue<string>() ?? Guid.NewGuid().ToString("N"),
            Type = node["type"]?.GetValue<string>() ?? PacketType.Error,
            Ts = node["ts"]?.GetValue<long>() ?? DateTimeOffset.UtcNow.ToUnixTimeMilliseconds(),
            Body = node["body"]?.AsObject() ?? new JsonObject()
        };
    }
}

File: android/app/src/main/java/io/conduit/protocol/Packet.kt
package io.conduit.protocol

import org.json.JSONObject
import java.util.UUID

/** Canonical packet `type` values. Must match PROTOCOL.md and the Windows side. */
object PacketType {
    const val IDENTITY = "identity"
    const val PAIR_REQUEST = "pair-request"
    const val PAIR_RESPONSE = "pair-response"
    const val PING = "ping"
    const val PONG = "pong"
    const val CLIPBOARD = "clipboard"
    const val FILE_OFFER = "file-offer"
    const val FILE_CHUNK = "file-chunk"
    const val FILE_COMPLETE = "file-complete"
    const val NOTIFICATION = "notification"
    const val NOTIFICATION_ACTION = "notification-action"
    const val MEDIA_STATE = "media-state"
    const val MEDIA_COMMAND = "media-command"
    const val REMOTE_COMMAND = "remote-command"
    const val BATTERY = "battery"
    const val DEVICE_STATUS = "device-status"
    const val SMS_LIST = "sms-list"
    const val SMS_SEND = "sms-send"
    const val ERROR = "error"
}

/**
 * The envelope every message travels in: { id, type, ts, body }.
 * See PROTOCOL.md §3.
 */
class Packet private constructor(
    val id: String,
    val type: String,
    val ts: Long,
    val body: JSONObject,
) {
    fun getString(key: String): String? = if (body.has(key)) body.optString(key) else null
    fun getLong(key: String, fallback: Long = 0L): Long = body.optLong(key, fallback)
    fun getInt(key: String, fallback: Int = 0): Int = body.optInt(key, fallback)
    fun getBool(key: String, fallback: Boolean = false): Boolean = body.optBoolean(key, fallback)
    fun getDouble(key: String): Double? = if (body.has(key)) body.optDouble(key) else null

    fun toJson(): String = JSONObject().apply {
        put("id", id)
        put("type", type)
        put("ts", ts)
        put("body", body)
    }.toString()

    companion object {
        fun create(type: String, build: (JSONObject.() -> Unit)? = null): Packet {
            val body = JSONObject()
            build?.invoke(body)
            return Packet(UUID.randomUUID().toString().replace("-", ""), type, System.currentTimeMillis(), body)
        }

        fun fromJson(json: String): Packet {
            val obj = JSONObject(json)
            return Packet(
                id = obj.optString("id", UUID.randomUUID().toString()),
                type = obj.optString("type", PacketType.ERROR),
                ts = obj.optLong("ts", System.currentTimeMillis()),
                body = obj.optJSONObject("body") ?: JSONObject(),
            )
        }
    }
}
